app/ghcup/BrickMain.hs

```diff
@@ -303,95 +483,96 @@ defaultAttributes no_color = attrMap
-      put (BrickState { appState = moveCursor 1 appState Down, .. })
-    (VtyEvent (Vty.EvResize _ _)) -> put st
-    (VtyEvent (Vty.EvKey Vty.KUp _)) ->
-      put BrickState{ appState = moveCursor 1 appState Up, .. }
-    (VtyEvent (Vty.EvKey Vty.KDown _)) ->
-      put BrickState{ appState = moveCursor 1 appState Down, .. }
-    (VtyEvent (Vty.EvKey key _)) ->
+    inner_event@(VtyEvent (Vty.EvKey key _)) ->
       case find (\(key', _, _) -> key' == key) (keyHandlers kb) of
-        Nothing -> put st
-        Just (_, _, handler) -> handler st
-    _ -> put st
-
-
-moveCursor :: Int -> BrickInternalState -> Direction -> BrickInternalState
-moveCursor steps ais@BrickInternalState{..} direction =
-  let newIx = if direction == Down then ix + steps else ix - steps
-  in  case clr !? newIx of
-        Just _  -> BrickInternalState { ix = newIx, .. }
-        Nothing -> ais
+        Nothing -> void $ Brick.zoom (toLensVL appState) $ handleGenericListEvent inner_event
+        Just (_, _, handler) -> handler
+    inner_event -> Brick.zoom (toLensVL appState) $ handleGenericListEvent inner_event
 
+eventHandler :: BrickEvent Name e -> EventM Name BrickState ()
+eventHandler ev = do
+  m <- use mode
+  case m of
+    Navigation -> navigationHandler ev
+    Tutorial   -> tutorialHandler ev
 
 -- | Suspend the current UI and run an IO action in terminal. If the
 -- IO action returns a Left value, then it's thrown as userError.
-withIOAction :: Ord n
-             => (BrickState
-                 -> (Int, ListResult)
-                 -> ReaderT AppState IO (Either String a))
-             -> BrickState
+withIOAction :: (Ord n, Eq n)
+             => ( (Int, ListResult) -> ReaderT AppState IO (Either String a))
              -> EventM n BrickState ()
-withIOAction action as = case listSelectedElement' (appState as) of
-  Nothing      -> put as
-  Just (ix, e) -> do
-    suspendAndResume $ do
-      settings <- readIORef settings'
-      flip runReaderT settings $ action as (ix, e) >>= \case
-        Left  err -> liftIO $ putStrLn ("Error: " <> err)
-        Right _   -> liftIO $ putStrLn "Success"
-      getAppData Nothing >>= \case
-        Right data' -> do
-          putStrLn "Press enter to continue"
-          _ <- getLine
-          pure (updateList data' as)
-        Left err -> throwIO $ userError err
+withIOAction action = do
+  as <- Brick.get
+  case sectionListSelectedElement (view appState as) of
+    Nothing      -> pure ()
+    Just (curr_ix, e) -> do
+      Brick.suspendAndResume $ do
+        settings <- readIORef settings'
+        flip runReaderT settings $ action (curr_ix, e) >>= \case
+          Left  err -> liftIO $ putStrLn ("Error: " <> err)
+          Right _   -> liftIO $ putStrLn "Success"
+        getAppData Nothing >>= \case
+          Right data' -> do
+            putStrLn "Press enter to continue"
+            _ <- getLine
+            pure (updateList data' as)
+          Left err -> throwIO $ userError err
```

Comment:
I've tried to make this part of the TUI, instead of `SuspendAndResume`. You can find a decent attempt [here](https://github.com/lsmor/ghcup-hs/tree/log-event). It isn't difficult, as it is only to extent the `Mode` type and to add a custom event channel, but there are some difficulties

Reply:
Yes, the logging and receiving the process output is the main issue. It's tricky with the current code.

Reply:
well the logging isn't a problem. It is relativaly easy to deviate the logging part to a Brick channel so It it can be used as a widget. The main problem is the output of external processes, for example the call to `curl` prints directly on top of the TUI... I don't think this feature will be implemented in this PR

Reply:
> The main problem is the output of external processes, for example the call to curl prints directly on top of the TUI... I don't think this feature will be implemented in this PR

That is what I meant with "logging". On the linux side we use low-level handling of stdout/stderr file descriptors... they could be exposed and used as callback style thing:

https://github.com/haskell/ghcup-hs/blob/9d8d6e329356bd9a9a27dd4628520284d7d97b2e/lib/GHCup/Prelude/Process/Posix.hs#L110-L115

On windows side we use the process package fake pipes:

https://github.com/haskell/ghcup-hs/blob/9d8d6e329356bd9a9a27dd4628520284d7d97b2e/lib/GHCup/Prelude/Process/Windows.hs#L161-L184

So a solution would probably have to write a wrapper around both that exposes the necessary parts and a callback that listens to messages on the pipe/Fd and reacts to it.